java/ru/ifmo/ctddev/kuplkri/arrayset/ArraySet.java

```diff
@@ -0,0 +1,403 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.NavigableSet;
+import java.util.NoSuchElementException;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+public class ArraySet<E> extends AbstractSet<E> implements NavigableSet<E> {
+
+    private E[] elements;
+    private int size;
+    private Comparator<? super E> customComparator;
+
+
+    @SuppressWarnings({"All"})
+    public ArraySet() {
+        elements = (E[]) new Object[0];
+    }
+
+    public ArraySet(Collection<E> values) {
+        createSortedArraySet(values);
+    }
+
+    public ArraySet(Collection<E> values, Comparator<? super E> customComparator) {
+        this.customComparator = customComparator;
+        createSortedArraySet(values);
+    }
+
+    @SuppressWarnings({"All"})
+    private void createSortedArraySet(final Collection<E> values) {
+        NavigableSet<E> arrayToSet = new TreeSet<>(customComparator);
+        arrayToSet.addAll(values);
+        elements = (E[]) arrayToSet.toArray();
+
+        this.size = elements.length;
+    }
+
+    @Override
+    public E lower(final E e) {
+        int ind = lowerIndex(e);
+        return ind == -1 ? null : elements[ind];
+    }
+
+    @Override
+    public E floor(final E e) {
+        int ind = floorIndex(e);
+        return ind == -1 ? null : elements[ind];
+    }
+
+    @Override
+    public E ceiling(final E e) {
+        int ind = ceilingIndex(e);
+        return ind == -1 ? null : elements[ind];
+    }
+
+    @Override
+    public E higher(final E e) {
+        int ind = higherIndex(e);
+        return ind == -1 ? null : elements[ind];
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @SuppressWarnings("ReturnOfInnerClass")
+    @Override
+    public Iterator<E> iterator() {
+        return new ArraySetIterator<>();
+    }
+
+    @Override
+    public NavigableSet<E> descendingSet() {
+        return new ArraySet<>(Arrays.asList(Arrays.copyOf(elements, size)), customComparator.reversed());
+    }
+
+    @Override
+    public Iterator<E> descendingIterator() {
+        return descendingSet().iterator();
+    }
+
+    @Override
+    public NavigableSet<E> subSet(final E fromElement, final boolean fromInclusive,
+                                  final E toElement, final boolean toInclusive) {
+        if (size == 0) {
+            return new ArraySet<>(Collections.emptyList(), customComparator);
+        }
+
+        int fromInd = fromInclusive ? ceilingIndex(fromElement) : higherIndex(fromElement);
+        int toInd = toInclusive ? floorIndex(toElement) : lowerIndex(toElement);
+
+        if (fromInd == -1 || toInd == -1 || toInd < fromInd) {
+            return new ArraySet<>(Collections.emptyList(), customComparator);
+        }
+
+        return new ArraySet<>(Arrays.asList(Arrays.copyOfRange(elements, fromInd, toInd + 1)), customComparator);
```

Comment:
look - you already know that `elements` are sorted. `ArraySet(Collection<E> values, Comparator<? super E> customComparator)` calls `createSortedArraySet(final Collection<E> values)`, which makes sorting again. You don't need sorting at this point. Please create one more private constructor `ArraySet(E[] values, Comparator<? super E> customComparator)`which just copies array part. This will reduce complexity from `O(n log n)` to `O(n)`